proposals/0000-net-storage-reads-improvements.md

```diff
@@ -0,0 +1,105 @@
+
+Unfortunately, this network protocol is suffering from some issues:
+
+- It is not possible for the querier to check whether a key exists in the storage of the chain except by querying the value of that key. The reply will thus include the value of the key, only for that value to be discarded by the querier that isn't interested by it. This is a waste of bandwidth.
+- It is not possible for the querier to know whether a value in the storage of the chain has been modified between two blocks except by querying this value for both blocks and comparing them. Only a few storage values get modified in a block, and thus most of the time the comparison will be equal. This leads to a waste of bandwidth as the values have to be transferred.
+- While it is possible to ask for multiple specific storage keys at the same time, it is not possible to ask for a list of keys that start with a certain prefix. Due to the way FRAME works, storage keys are grouped by "prefix", for example all account balances start with the same prefix. It is thus a common necessity for a light client to obtain the list of all keys (and possibly their values) that start with a specific prefix. This is currently not possible except by performing multiple queries serially that "walk down" the trie.
+
+Once Polkadot and Kusama will have transitioned to `state_version = 1`, which modifies the format of the trie entries, it will be possible to generate Merkle proofs that contain only the hashes of values in the storage. Thanks to this, it is already possible to prove the existence of a key without sending its entire value (only its hash), or to prove that a value has changed or not between two blocks (by sending just their hashes).
+Thus, the only reason why aforementioned issues exist is because the existing networking messages don't give the possibility for the querier to query this. This is what this proposal aims at fixing.
+
+## Detailed Solution design
+The protobuf schema of the networking protocol can be found here: https://github.com/paritytech/substrate/blob/5b6519a7ff4a2d3cc424d78bc4830688f3b184c0/client/network/light/src/schema/light.v1.proto
+
+The proposal is to modify this protocol in this way:
+
+```diff
+@@ -11,6 +11,7 @@ message Request {
+                RemoteReadRequest remote_read_request = 2;
+                RemoteReadChildRequest remote_read_child_request = 4;
+                // Note: ids 3 and 5 were used in the past. It would be preferable to not re-use them.
++               RemoteReadRequestV2 remote_read_request_v2 = 6;
+        }
+ }
+ 
+@@ -48,6 +49,21 @@ message RemoteReadRequest {
+        repeated bytes keys = 3;
+ }
+ 
++message RemoteReadRequestV2 {
++       required bytes block = 2;
++       optional ChildTrieInfo child_trie_info = 3;  // Read from the main trie if missing.
++       repeated Key keys = 6;
```

Comment:
Why are these indices not 1-3? Is there any reason behind this?

Reply:
Good question, I haven't actually noticed that. I copy-pasted `RemoteReadChildRequest` and it uses 6 as well, but I don't know why it does.
As long as they're inferior to 128, this will always use 1 byte anyway, no matter the value.

Reply:
> As long as they're inferior to 128, this will always use 1 byte anyway, no matter the value.

https://protobuf.dev/programming-guides/proto3/#assigning-field-numbers 0-15 actually only take 1 byte (read this to understand why not 1-3).

> Good question, I haven't actually noticed that. I copy-pasted `RemoteReadChildRequest` and it uses 6 as well, but I don't know why it does.

If we introduce some new answer here, could we please change to 1 - 3 as indices? Otherwise people may assume that there are something was removed from this message over time.

Reply:
> https://protobuf.dev/programming-guides/proto3/#assigning-field-numbers 0-15 actually only take 1 byte (read this to understand why not 1-3).

Ah sorry, I forgot about the type thing